src/context: priorize bootchooser to select bootslot (for custom bootloader backend)

gportay opened the pull request:
The commit 9af0613c7527ebadc855c1a49a4d10757150194a introduced the new bootchooser API r_boot_get_current_bootname() to allow the backend to determine the current bootslot.

Since then, the bootchooser is able determine the bootslot if it cannot be guessed from the kernel cmdline; it is used by the custom backend only at that moment.

This gives the priority to the bootchooser over the kernel cmdline.

<!--
Thank you for your pull request!

If this is your first pull request for RAUC, please read:
https://rauc.readthedocs.io/en/latest/contributing.html

Please describe what it changes, e.g. fixes a bug (and how), adds a feature, fixes documentation…

If you add a feature, please answer these questions:
- What do you use the feature for?
- How does RAUC benefit from the feature?
- How did you verify the feature works?
- If hardware is needed for the feature, which hardware is supported and which
  hardware did you test with?

Please also allow edits by maintainers, so we can apply minor fixes directly.
-->

<!--
In case your PR fixes an issue, please reference it in the next line, i.e.
Fixes: #[insert number without brackets here]
-->

ejoerns commented:
Sounds reasonable to me to give the custom backend priority for setting the booted slot.

`get-current` remains optional since missing output or any failure in get-current (e.g. due to strict input checking) will not lead to an error and just fall back to the  `root=`-based  bootname.

I guess it would make sense to clarify the detection order in the description in https://rauc.readthedocs.io/en/latest/integration.html#identification-via-custom-backend

jluebbe commented:
@ejoerns I've not checked the code yet, but we need to ensure that external and NFS override results from the bootloader backend.

ejoerns commented:
@jluebbe Indeed, that's one corner case I overlooked. Thus maybe it makes sense to just move the call to `r_boot_get_current_bootname()` into `get_cmdline_bootname()` (and rename this?).

It's also a bit inconsistent currently that we check `rauc.external` first, then `rauc.slot` and `bootchooser.active`, THEN `root=/dev/nfs`, and finally other `root=` options.
I'd suggest we check `rauc.external` and `root=/dev/nfs` first, then handle the custom backend, and then the rest.

oli-ben commented:
> @jluebbe Indeed, that's one corner case I overlooked. Thus maybe it makes sense to just move the call to `r_boot_get_current_bootname()` into `get_cmdline_bootname()` (and rename this?).
> 
> It's also a bit inconsistent currently that we check `rauc.external` first, then `rauc.slot` and `bootchooser.active`, THEN `root=/dev/nfs`, and finally other `root=` options. I'd suggest we check `rauc.external` and `root=/dev/nfs` first, then handle the custom backend, and then the rest.

I ran into the same issue as @gportay, I think this would be a great path forward!

gportay commented:
@jluebbe, @ejoerns, I have split the `get_cmdline_bootname()` logic in two. I have moved the explicit `rauc slot` before calling the bootloader backend (and the nfs case). I have kept the rauc slot guessing from the root device after.

Also, I am thinking about moving the `barebox` thing in the bootloader backend by adding the cmdline in parameter to the `r_boot_get_current_bootname(..., cmdline, ...)`.

Is that looking good to you?

jluebbe commented:
I've created #1712 to make the refactoring easier to review. What do you think, @gportay?

gportay commented:
> I've created #1712 to make the refactoring easier to review. What do you think, @gportay?

I saw, let's close that one.
